syntax = "proto3";

package matrix_service;

service MatrixService {
  rpc SendRow(MatrixRow) returns (RowResponse);
  rpc Query(QueryRequest) returns (QueryResponse);
}

message MatrixRow {
  int32 client_id = 1;
  int32 row_index = 2;
  repeated double values = 3;
}

message RowResponse {
  bool success = 1;
  string message = 2;
  int32 rows_received = 3;
  int32 total_expected_rows = 4;
}

message QueryRequest {
  int32 client_id = 1;
  int32 query_type = 2; 
}

message QueryResponse {
  bool success = 1;
  string message = 2;
  oneof result {
    int32 row_count = 3;
    int32 rank = 4;
    double determinant = 5;
    string waiting_status = 6;
    bool has_min_rank = 7;  
    bool has_min_determinant = 8;
  }
}
